sdk/src/impl/whirlpool-impl.ts

```diff
@@ -328,65 +335,117 @@ export class WhirlpoolImpl implements Whirlpool {
     destinationWallet: PublicKey,
     positionWallet: PublicKey,
     payerKey: PublicKey
-  ): Promise<TransactionBuilder> {
+  ): Promise<{ ataTx?: TransactionBuilder; closeTx: TransactionBuilder }> {
     const position = await this.ctx.fetcher.getPosition(positionAddress, true);
     if (!position) {
       throw new Error(`Position not found: ${positionAddress.toBase58()}`);
     }
+
     const whirlpool = this.data;
 
     invariant(
       position.whirlpool.equals(this.address),
       `Position ${positionAddress.toBase58()} is not a position for Whirlpool ${this.address.toBase58()}`
     );
 
+    const positionTokenAccount = await deriveATA(positionWallet, position.positionMint);
+
+    const ataTxBuilder = new TransactionBuilder(
+      this.ctx.provider.connection,
+      this.ctx.provider.wallet
+    );
+
+    const txBuilder = new TransactionBuilder(
+      this.ctx.provider.connection,
+      this.ctx.provider.wallet
+    );
+
     const tickArrayLower = PDAUtil.getTickArrayFromTickIndex(
       position.tickLowerIndex,
       whirlpool.tickSpacing,
       position.whirlpool,
       this.ctx.program.programId
     ).publicKey;
+
     const tickArrayUpper = PDAUtil.getTickArrayFromTickIndex(
       position.tickUpperIndex,
       whirlpool.tickSpacing,
       position.whirlpool,
       this.ctx.program.programId
     ).publicKey;
 
-    const positionTokenAccount = await deriveATA(positionWallet, position.positionMint);
+    const affiliatedMints = getTokenMintsFromWhirlpools([whirlpool]);
+    const accountExemption = await this.ctx.fetcher.getAccountRentExempt();
+    const { ataTokenAddresses, resolveAtaIxs } = await resolveAtaForMints(this.ctx, {
+      mints: affiliatedMints,
+      accountExemption,
+      receiver: destinationWallet,
+      payer: payerKey,
+    });
 
-    const txBuilder = new TransactionBuilder(
-      this.ctx.provider.connection,
-      this.ctx.provider.wallet
+    ataTxBuilder.addInstructions(resolveAtaIxs);
+
+    const tickArrayLowerData = await this.ctx.fetcher.getTickArray(tickArrayLower, true);
+    const tickArrayUpperData = await this.ctx.fetcher.getTickArray(
+      tickArrayUpper,
+      !tickArrayUpper.equals(tickArrayLower)
```

Comment:
`once` https://github.com/orca-so/whirlpools/pull/63 goes in, you should be able to skip all this and fetch the whirlpoolData & tick array data from the PositionImpl object

Reply:
ah whoopies. What I meant was
1. For the tick-array fetching part, you can use `getTickArrayDataForPosition` which can simplify this a bit for you
2. Once you have the positionimpl object, you can call getLowerTickData / getUpperTickData.